File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-7.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        keys = defaultdict(int)
        if s == '':
            return 0
        ans = 1
        left, right = 0, 0
        while (left < len(s) and right < len(s)):
            if s[right] in keys:
                ans = max(ans, right-left)
                keys.pop(s[left], None)
                left += 1
            else:
                keys[s[right]] = right
                right += 1

        ans = max(ans, right - left)
        return ans
            

File: Data Structures & Algorithms/permutation-string/submission-7.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if(s1.size() > s2.size()) return false;
        
        string n = "00000000000000000000000000", w = "00000000000000000000000000";
        
        for(auto &it: s1) n[it-'a']++;

        for(int i = 0; i < s2.size(); ++i) {
            w[s2[i]-'a']++;
            if (i >= s1.size()) w[s2[i-s1.size()] - 'a']--;
            if(w==n) return true;
        } 
        return false;
    }
};

